### PR TITLE
[inetstack] Move PacketBuf to Layer 2

### PR DESCRIPTION
**Pull request by iyzhang**
This PR moves the PacketBuf data structure to Layer 2 and removes many of the dependencies on abstractions from Layer2 (e.g., local MAC address) into the L2 module. The goal is to keep moving until we get rid of the PacketBuf data structure altogether.

### CODE REVIEW

**Review by ppenna — APPROVED** [review]
LGTM

File: src/rust/inetstack/protocols/tcp/established/ctrlblk.rs [hunk 1]
```diff
@@ -858,7 +854,11 @@ impl SharedControlBlock {
         };
 
         // Call the runtime to send the segment.
```
Comment (anandbonde) [inline src/rust/inetstack/protocols/tcp/established/ctrlblk.rs]: This should be adjusted.

File: src/rust/inetstack/protocols/tcp/passive_open.rs [hunk 2]
```diff
@@ -334,7 +330,11 @@ impl SharedPassiveSocket {
         };
 
         // Send it.
```
Comment (anandbonde) [inline src/rust/inetstack/protocols/tcp/passive_open.rs]: This can be removed.

File: src/rust/inetstack/protocols/tcp/tests/simulator.rs [hunk 3]
```diff
@@ -850,36 +850,48 @@ impl Simulation {
...
             Some(Self::cook_buffer(tcp_packet.seqnum.win as usize, None))
         } else {
             None
         };
+        let mut segment: TcpSegment =
+            TcpSegment::new(ipv4_hdr, tcp_hdr, data, false).expect("Should be able to create TCP segment");
+        let mut pkt: DemiBuffer = segment.take_body().unwrap();
 
-        TcpSegment::new(ethernet2_hdr, ipv4_hdr, tcp_hdr, data, false).unwrap()
+        // Actually prepend L2 header.
```
Comment (anandbonde) [inline src/rust/inetstack/protocols/tcp/tests/simulator.rs:866]: This is identical for tcp and udp, we should extract it into a separate fn if possible.

**Pull request merged**